### DISCUSSION TITLE
Authenticating remotely with CLI link without window prompt

### DISCUSSION THREAD

**Discussion by dantaeusb** [yuezk/GlobalProtect-openconnect#453] [Q&A] [answered]
I want to use GlobalProtect on a IDE thin client host machine, and instead of opening a window on that machine, I'd like to receive the authentication URL in the CLI and use it on a client to authenticate. Is it possible? Machines are in the same network under NAT so I expect them to not cause any conflicts.

**Comment by yuezk** [comment]
Hi @dantaeusb,

You may try this if I understand your requirements correctly:

1. Use `gpauth` to authenticate and obtain the auth cookie outside. `gpauth <portal> > cookie.txt`
2. Copy `cookie.txt` to your thin client host machine and run `cat cookie.txt | sudo gpclient connect <portal> --cookie-on-stdin`

**Reply by dantaeusb** [reply]
Thank you, I think I got the idea. 
Running `gpauth` on host opens the window on the host machine, which I'd like to disconnect from the display. Unfortunately, the client is macOS, but I would try to build `gpauth` for macOS, maybe it'll work, and then I'll just send cookie over ssh. Probably that's the best way to do that – as even if I had had the URI, I'd still needed to extract the cookie and transfer it to the host somehow.

**Reply by dantaeusb** [reply]
Ok that would require me to do webview on something like DarwinWerKit, but hopefully that shouldn't be too much of a problem.

**Reply by yuezk** [reply]
Currently, the `gpauth` doesn't support macOS.

**Reply by yuezk** [reply, marked as answer]
> Currently, the `gpauth` doesn't support macOS.

`gpauth` now supports macOS. You can build the `gpauth` binary with `cargo build -p gpauth` on macOS

**Reply by dantaeusb** [reply]
That works flawlessly, a little shell script and I can authenticate the host easily – the only extra step I did is allow running client without sudo to avoid passing password.
A million thanks!

**Reply by mlanner** [reply]
@dantaeusb Do you have a sample script you could provide? I tried the same building my own script, but couldn't quite get it to work. I successfully used `gpauth` locally, used **scp** to transfer the cookie file to the remote machine and then tried to connect by issuing the command `cat cookie.txt | sudo gpclient connect <portal> --cookie-on-stdin` over **ssh**, but it fails. Any suggestions or a sample script would be greatly appreciated.

**Reply by dantaeusb** [reply]
@mlanner first what I did to avoid passing password over ssh or using superuser is to add to sudoers sudo visudo the option to run gpclient without auth, like
`<username> ALL = NOPASSWD: /usr/bin/gpclient`
Then I use the following script

```bash
#!/bin/bash
set -e

DATE=$(date +"%Y-%m-%d-%H-%M")
FILENAME="cookie_${DATE}.txt"
./gpauth <portal> > $FILENAME
scp ./$FILENAME <user>@<host>:~/Tools/$FILENAME
ssh <user>@<host> "cat ~/Tools/$FILENAME | sudo -E gpclient connect <portal> --cookie-on-stdin"
rm $FILENAME
```

I recommend also extending it with screen or tmux so you don't have to keep terminal but have access to what's happening inside like:
`ssh user@ip "tmux new-session -d -s VPN 'cat ~/Tools/$FILENAME | sudo -E gpclient connect <portal> --cookie-on-stdin'"` (`tmux`/`screen` should answer the https://github.com/yuezk/GlobalProtect-openconnect/discussions/462 as well)
And deleting file from host later. You can probably skip file transfer altogether, but while I was writing the script it was useful for retries and debugs.

**Reply by mlanner** [reply]
@dantaeusb Thx! That's very similar to what I did, and I did actually try to run it in `tmux`, but it somehow didn't work. My test Linux account has full passwordless sudo, so I think the sudoers part is sorted. I'll take your version and see if I can get it to work without `tmux` first.

**Reply by mlanner** [reply]
Well, it turns out I think my `gpauth` call probably doesn't work as expected and I'm suspecting that as a result the cookie.txt file is likely missing some info:

```
{"success":{"portalUserauthcookie":"","preloginCookie":"theCookie","token":null,"username":"testuser"}}
```

Specifically the `"token":null` part.

I also see a bunch of other info about "No auth data ...", like:

```
[2025-01-17T03:28:28Z INFO  auth::webview::webview_auth] Finished loading page: https://************/login/token/redirect?stateToken=************
[2025-01-17T03:28:28Z INFO  auth::webview::webview_auth] No auth data found in Headers, it may not be the /SAML20/SP/ACS endpoint
[2025-01-17T03:28:28Z INFO  auth::webview::auth_messenger] Read auth data from html failed: No auth data found, extracting gpcallback...
[2025-01-17T03:28:28Z INFO  auth::webview::webview_auth] No auth data found in Body, it may not be the /SAML20/SP/ACS endpoint
```

So, once I scp the file to the other machine, I'm thinking it doesn't have the proper cookie data to operate on to start the VPN connection.

Maybe I'm wrong. Just making some guesses ... 

Thx!

**Reply by yuezk** [reply]
The `token` can be `null`, and the `No auth data...` is because the loaded page is not the token response page. Under the hood, `gpauth` keeps inspecting the requests until it finds the auth cookie from the response.

Can you post your full command used?

**Reply by mlanner** [reply]
@yuezk, this is my adapted script:

```
#!/bin/bash
set -e

VPN_PORTAL="gp.example.com"
VPN_HOST="192.168.1.10"
VPN_HOST_USER="testuser"
TMUX_SESSION="gp-vpn"
DATE=$( date +"%Y%m%d-%H%M" )
COOKIE_DIR="${HOME}/tmp"
COOKIE_FILE="${COOKIE_DIR}/gpclient_cookie_${DATE}.txt"

# Get cookie
/usr/bin/gpauth ${VPN_PORTAL} > ${COOKIE_FILE}

# Transfer cookie to VPN host
scp ${COOKIE_FILE} ${VPN_HOST_USER}@${VPN_HOST}:/home/${VPN_HOST_USER}/tmp/

# Start VPN via SSH command on VPN host
ssh ${VPN_HOST_USER}@${VPN_HOST} "tmux new-session -d -s ${TMUX_SESSION} cat /home/${VPN_HOST_USER}/tmp/${COOKIE_FILE} | sudo -E gpclient connect ${VPN_PORTAL} --cookie-on-stdin'"

# Clean up local cookie file
rm ${COOKIE_FILE}
```

Thx!

**Reply by yuezk** [reply]
@mlanner The script looks good. Can you provide the logs? Thanks.

**Comment by buldezir** [comment]
simplier script, without file copying and without tmux:
```
function vm_vpn_connect() {
  _GPPORTAL="gp.example.com"
  _GPCOOKIE=$(gpauth -q $_GPPORTAL)
  ssh user@host "sudo pkill -SIGINT gpclient; sleep 2; echo '$_GPCOOKIE' | sudo -Eb gpclient connect --cookie-on-stdin $_GPPORTAL >/dev/null 2>&1 &"
}
```
